#+TITLE: Extending =ensembldb=: MySQL backend and protein annotations
#+AUTHOR: Johannes Rainer (EURAC research, Italy), Sebastian Gibb, Laurent Gatto (CPU Cambridge, UK)
#+EMAIL: johannes.rainer@eurac.edu
#+DATE: December 7, 2016
#+LATEX_HEADER: \usepackage[backend=bibtex,style=chem-rsc,hyperref=true]{biblatex}
#+LATEX_HEADER: \usepackage{parskip}
#+LATEX_HEADER: \addbibresource{~/Documents/Unison/bib/references.bib}
#+LATEX_HEADER: \usepackage{inconsolata}
#+LATEX_HEADER: \definecolor{lightgrey}{HTML}{F0F0F0}
#+LATEX_HEADER: \definecolor{solarizedlightbg}{HTML}{FCF4DC}
#+LATEX_HEADER: \makeatletter
#+LATEX_HEADER: \patchcmd{\@verbatim}
#+LATEX_HEADER:   {\verbatim@font}
#+LATEX_HEADER:   {\verbatim@font\scriptsize}
#+LATEX_HEADER:   {}{}
#+LATEX_HEADER: \makeatother

#+PROPERTY: header-args :exports both
#+PROPERTY: header-args :noweb yes
#+PROPERTY: header-args :results output verbatim
#+PROPERTY: header-args :tangle yes
#+PROPERTY: header-args:R :session *R_EuroBioC2016*

#+LATEX_CLASS: beamer
#+LATEX_CLASS_OPTIONS: [presentation,smaller]
#+BEAMER_THEME: default
#+BEAMER_COLOR_THEME: eurac
#+BEAMER_INNER_THEME: circles
#+COLUMNS: %40ITEM %10BEAMER_env(Env) %9BEAMER_envargs(Env Args) %4BEAMER_col(Col) %10BEAMER_extra(Extra)
#+OPTIONS: toc:nil
#+OPTIONS: H:2
#+OPTIONS: email:nil
#+OPTIONS: author:t


* Extending =ensembldb=: MySQL backend and protein annotations

** Introduction

+ =ensembldb=: retrieve gene & transcript annotations.
+ =ensembldb= package defines the =EnsDb= class:
  - Wrapper to access annotations from an SQLite database.
  - Same functionality than the =GenomicFeatures= package (=TxDb= object).
  - Filter framework to enable specific and fast queries.
  - *NEW*: /MySQL/ backend support.
  - *NEW*: provide protein annotations.


** Basic usage

+ Available methods to extract data: =genes=, =transcripts=, =transcriptsBy=, =exons=,
  =exonsBy=, =cdsBy=, =fiveUTRsByTranscripts=, =threeUTRsByTranscripts=, =proteins= (*NEW*).

+ _Example_: get all lincRNA genes encoded on chromosome Y.

  #+BEGIN_SRC R :exports both :results silent
    ## Load an EnsDb package matching Ensembl version 86
    library(EnsDb.Hsapiens.v86)
    edb <- EnsDb.Hsapiens.v86

    ## Retrieve all lincRNAs encoded on chromosome Y.
    ## Create the filter objects
    sf <- SeqnameFilter("Y")
    gbf <- GenebiotypeFilter("lincRNA")

  #+END_SRC

** Basic usage

+ _Example:_ (continued)

  #+BEGIN_SRC R :exports both :results output pp
    ## Retrieve the data.
    genes(edb, filter = list(sf, gbf))
  #+END_SRC

** Available filters

- For genes: =GeneidFilter=, =GenenameFilter=, =EntrezidFilter=, =GenebiotypeFilter=,
  (=SymbolFilter=).
- For transcripts: =TxidFilter=, =TxbiotypeFilter=.
- For exons: =ExonidFilter=, =ExonrankFilter=.
- *NEW*: for proteins: =ProteinidFilter=, =UniprotidFilter=, =UniprotdbFilter=,
  =UniprotmappingtypeFilter=, =ProtdomidFilter=.
- Based on chromosomal coordinates: =SeqnameFilter=, =SeqstrandFilter=,
  =SeqstartFilter=, =SeqendFilter=, =GRangesFilter=: condition can be /within/ or
  /overlapping/.

- Multiple filters are combined with a logical /AND/.
- Each filter supports 1:n values, /=/, /!=/ and also a /like/ condition.


** Building annotation databases

+ *Option A)*: from GTF/GFF files or =AnnotationHub=.
+ _Example_: create an =EnsDb= using =AnnotationHub=.

  #+BEGIN_SRC R :results silent :exports code :eval never
    library(AnnotationHub)
    ah <- AnnotationHub()
    ## Query for available Ensembl gtf files for release 83.
    query(ah, pattern=c("ensembl", "release-83", "gtf"))

    ## Select one; in this case: Anolis carolinensis (lizard)
    edbSql83 <- ensDbFromAH(ah=ah["AH50353"])
    ## BUT: DB lacks NCBI Entrezgene IDs and protein annotation.

    ## Load the database.
    db <- EnsDb(edbSql83)

    ## Optional, make a package.
    makeEnsembldbPackage(ensdb=edbSql83, version="1.0.0", author="J Rainer",
                         maintainer="Johannes Rainer <johannes.rainer@eurac.edu>")
  #+END_SRC

+ *Option B)* (preferred): using the Ensembl Perl API:
  - =fetchTablesFromEnsembl= and =makeEnsemblSQLiteFromTables=.
  - Fetches also *protein annotations*.
  - Requirements: Perl, Bioperl, Ensembl Perl API.


** MySQL backend

+ _Example_: =listEnsDb= list all available databases, =useMySQL= to insert an =EnsDb=
  to a MySQL server.

  #+BEGIN_SRC R :results output pp :exports both
    library(RMySQL)
    dbc <- dbConnect(MySQL(), host = "localhost", user = "anonuser", pass = "")

    ## list all available EnsDb databases.
    listEnsDbs(dbc)
  #+END_SRC

  #+BEGIN_SRC R :results output pp
    ## Connect to a database.
    dbc <- dbConnect(MySQL(), host = "localhost", user = "anonuser", pass = "",
                     dbname = "ensdb_dmelanogaster_v86")
    edb_2 <- EnsDb(dbc)

    ## To insert an EnsDb to a MySQL: useMySQL
    db_my <- useMySQL(edb, host = "localhost", user = "anonuser", pass = "")
  #+END_SRC

+ Enables a central, MySQL-based annotation server.


** Protein data: fetch protein data from an =EnsDb=

+ _Example_: add protein columns to the =columns= parameter.

  #+BEGIN_SRC R :results output pp :exports both
    ## Get all genes with a C2H2 Zinc finger domain and
    ## return all of their Uniprot IDs
    pfam <- ProtdomidFilter("PF13912")
    genes(edb, filter = pfam, return.type = "DataFrame",
  	columns = c("gene_name", "uniprot_id"))
  #+END_SRC

** Protein data: fetch protein data from an =EnsDb=

+ _Example_: use =proteins= method to specifically fetch protein data.

  #+BEGIN_SRC R :results output pp :exports both
    ## Return the protein annotation as a AAStringSet:
    prts <- proteins(edb, filter = GenenameFilter("ZBTB16"),
                     columns = c("tx_id", "tx_biotype"),
                     return.type = "AAStringSet")
    prts
  #+END_SRC
  #+BEGIN_SRC R :results output pp :exports both
    ## Additional columns are available as mcols:
    mcols(prts)
  #+END_SRC

** Protein data: use =ensembldb= with =Pbase=

+ =Pbase=: (Laurent Gatto and Sebastian Gibb): provides classes and functions for
  the analysis of protein sequence data in proteomics experiments.

+ The =Proteins= object: container for proteins and peptide ranges within the AA
  sequences.

+ _Example_: fetch a =Proteins= object for all ZBTB16 proteins including their protein
  domains.

  #+BEGIN_SRC R :results output pp :exports both
    ## load Pbase - we need the "ensembldb" branch.
    library(Pbase)

    ## Fetch proteins including protein domains for ZBTB16
    prts <- Proteins(edb, filter = GenenameFilter("ZBTB16"))

    ## Amino acid sequence:
    aa(prts)
  #+END_SRC

** Protein data: use =ensembldb= with =Pbase=
+ _Example_: fetch a =Proteins= object for all ZBTB16 proteins including their protein
  domains (continued).
  #+BEGIN_SRC R :results output pp :exports both
    ## Peptide features:
    pranges(prts)
  #+END_SRC

** Protein data: use =ensembldb= with =Pbase=

+ _Example_: use =ensembldb= to map peptide features to the genome.

  #+NAME: map-plot
  #+BEGIN_SRC R :results output graphics :file gviz.pdf :width 8 :height 5
    ## Map all protein domains to the genome
    gen_map <- mapToGenome(prts, edb)

    ## Plot the results for the first protein (transcript)
    txid <- gen_map[[1]]$tx_id
    ## Get the gene region track for the first transcript
    tx <- getGeneRegionTrackForGviz(edb, filter = TxidFilter(txid))

    ## Add a protein ID column
    map_1 <- gen_map[[1]]
    map_1$id <- names(map_1)

    ## Plot using Gviz
    library(Gviz)
    plotTracks(list(GenomeAxisTrack(),
                    GeneRegionTrack(tx, name = "tx"),
                    AnnotationTrack(map_1, groupAnnotation = "id",
                                    just.group = "above",
                                    name = "Protein domains")),
               transcriptAnnotation = "transcript")
  #+END_SRC

** Protein data: use =ensembldb= with =Pbase=

#+ATTR_LATEX: :center :placement [H] :width 11cm
[[file:gviz.pdf]]


** Things not covered

+ =ensembldb= provides full =AnnotationDbi= support.
+ _Example_: use =AnnotationDbi='s =select= method to fetch annotations: =keys= can be a
  character list of IDs or a list of filter objects.

  #+BEGIN_SRC R :results output pp :exports both
    ## Get all data for the gene SKA2
    Res <- select(edb, keys="SKA2", keytype="GENENAME")
    head(Res, n=3)
  #+END_SRC

** Things not covered

+ Easy integration of UCSC and Ensembl annotations: use =seqlevelsStyle= to
  change chromosome naming scheme in =EnsDb=.
+ _Example_: How to integrate Ensembl based annotation with UCSC data?

  #+BEGIN_SRC R :results output pp :exports both
    ## Get chromosome names, they are "Ensembl-formatted"
    head(seqlevels(edb))
  #+END_SRC
  #+BEGIN_SRC R :results output pp :exports both
    ## Get genes on chromosome Y, UCSC style.
    genes(edb, filter=SeqnameFilter("chrY"))
  #+END_SRC
  #+BEGIN_SRC R :results output pp :exports both
    ## Solution: change the chromosome naming style:
    seqlevelsStyle(edb) <- "UCSC"
    ## Get chromosome names
    head(seqlevels(edb))
  #+END_SRC
  #+BEGIN_SRC R :results output pp
    genes(edb, filter=SeqnameFilter("chrY"))


    ## Use case:
    ## Get mRNA sequences for SKA2 using BSgenome.
    library(BSgenome.Hsapiens.UCSC.hg38)  ## <- UCSC based

    ## Get exons by transcript
    ska2tx <- exonsBy(edb, by="tx", filter=GenenameFilter("SKA2"))

    ## Use GenomicFeatures' extractTranscriptSeqs
    head(extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg38, ska2tx))


    ## Alternative (preferred) way:
    seqlevelsStyle(edb) <- "Ensembl"
    ## Using AnnotationHub:
    ## Get the genomic fasta file matching the package's genome version:
    faf <- getGenomeFaFile(edb)
    extractTranscriptSeqs(faf, exonsBy(edb, by="tx",
                                       filter=GenenameFilter("SKA2")))
  #+END_SRC


** Finally


*Thank you for your attention!*



https://github.com/jotsetung/EuroBioC2016-ensembldb.git


#+BEGIN_SRC R :results silent :exports none
  ## Clean up.
  dbDisconnect(dbc)

  rm(list = ls())
  gc()
  gc()
#+END_SRC






















